File: devscout-agent/agent.py
#!/usr/bin/env python3
"""
DevScout — demand-side reference agent (developer tools / SaaS).

Flow: GET /ad-units → GET manifest → POST actions/report → optional sandbox → conversion.

Defaults to dry run (MB_DRY_RUN=1). Set MB_DRY_RUN=0 to report attribution.
Set MB_ALLOW_LIVE=1 and MB_SANDBOX_HOST_ALLOWLIST for partner sandbox calls only.
"""

from __future__ import annotations

import json
import os
import sys
import uuid
from urllib.parse import urlencode, urljoin

import requests

DEFAULT_BASE = "https://www.moltbillboard.com"
TARGET_INTENTS = {"software.purchase", "subscription.register"}


def env(name: str, default: str | None = None) -> str | None:
    value = os.environ.get(name)
    return value if value not in (None, "") else default


def divider(title: str) -> None:
    print(f"\n{'─' * 62}\n  {title}\n{'─' * 62}")


def discover_ad_units(base: str, topic: str, limit: int) -> list[dict]:
    params = urlencode({"topic": topic, "limit": limit, "surface": "api"})
    url = urljoin(base, f"/api/v1/ad-units?{params}")
    response = requests.get(url, timeout=15)
    response.raise_for_status()
    return response.json().get("adUnits") or []


def rank_unit(units: list[dict]) -> dict | None:
    def score(unit: dict) -> int:
        s = 0
        if unit.get("primaryIntent") in TARGET_INTENTS:
            s += 20
        if unit.get("actionEndpoint"):
            s += 5
        return s

    if not units:
        return None
    return max(units, key=score)


def pick_offer(manifest: dict) -> tuple[dict, dict]:
    placement = manifest.get("placement") or manifest.get("manifest", {}).get("placement", {})
    offers = placement.get("offers") or []
    if not offers:
        raise RuntimeError("Manifest has no offers")

    def offer_score(offer: dict) -> int:
        s = 0
        if offer.get("primaryIntent") in TARGET_INTENTS:
            s += 30
        if offer.get("isPrimary"):
            s += 8
        return s

    offer = max(offers, key=offer_score)
    action_id = (offer.get("attribution") or {}).get("actionId")
    if not action_id:
        raise RuntimeError("Offer missing actionId from manifest")
    return placement, offer


def main() -> None:
    base = env("MB_BASE", DEFAULT_BASE).rstrip("/")
    topic = env("MB_TOPIC", "developer tools")
    limit = int(env("MB_LIMIT", "5") or "5")
    dry_run = env("MB_DRY_RUN", "1") != "0"
    allow_live = env("MB_ALLOW_LIVE", "0") == "1"
    allowlist = {
        h.strip().lower()
        for h in (env("MB_SANDBOX_HOST_ALLOWLIST") or "").split(",")
        if h.strip()
    }

    divider("DevScout agent")
    print(f"  Base:     {base}")
    print(f"  Topic:    {topic}")
    print(f"  Dry run:  {dry_run}")

    units = discover_ad_units(base, topic, limit)
    if not units:
        sys.exit("No ad units for topic")

    unit = rank_unit(units)
    placement_id = unit["placementId"]
    print(f"  Ad unit:  {unit.get('title') or placement_id}")

    manifest_resp = requests.get(
        urljoin(base, f"/api/v1/placements/{placement_id}/manifest"),
        timeout=15,
    )
    manifest_resp.raise_for_status()
    envelope = manifest_resp.json()
    manifest = envelope.get("manifest") or envelope
    placement, offer = pick_offer(manifest)
    placement_id = placement.get("id") or placement_id
    offer_id = offer["offerId"]
    action_id = offer["attribution"]["actionId"]
    action_endpoint = offer.get("actionEndpoint") or unit.get("actionEndpoint")

    print(f"  Offer:    {offer_id}")
    print(f"  Action:   {action_id}")

    if dry_run:
        print("\n  Dry run complete (MB_DRY_RUN=1). Set MB_DRY_RUN=0 to report events.")
        return

    idem = f"devscout-{uuid.uuid4()}"
    report = requests.post(
        urljoin(base, "/api/v1/actions/report"),
        headers={"Content-Type": "application/json", "Idempotency-Key": idem},
        json={
            "actionId": action_id,
            "placementId": placement_id,
            "offerId": offer_id,
            "eventType": "offer_selected",
        },
        timeout=15,
    )
    report.raise_for_status()
    print(f"  offer_selected: {report.json().get('success')}")

    executed = False
    if allow_live and action_endpoint and allowlist:
        from urllib.parse import urlparse

        host = urlparse(action_endpoint).hostname or ""
        if host.lower() in allowlist:
            merchant = requests.post(
                action_endpoint,
                json={"requester": "devscout-agent", "sandbox_mode": True},
                timeout=20,
            )
            executed = merchant.status_code in (200, 201)
            print(f"  sandbox:  HTTP {merchant.status_code}")

    if executed or env("MB_REPORT_CONVERSION_WITHOUT_LIVE", "0") == "1":
        conv = requests.post(
            urljoin(base, "/api/v1/conversions/report"),
            json={
                "actionId": action_id,
                "placementId": placement_id,
                "offerId": offer_id,
                "conversionType": env("MB_CONVERSION_TYPE", "signup"),
            },
            timeout=15,
        )
        conv.raise_for_status()
        print(f"  conversion: {conv.json().get('success')}")

    print("\n  DevScout completed.")


if __name__ == "__main__":
    try:
        main()
    except Exception as exc:
        print(f"Error: {exc}", file=sys.stderr)
        sys.exit(1)

File: devscout-agent/agent.ts
/**
 * DevScout (TypeScript, no SDK) — ad-units → manifest → attribution.
 * Requires Node 18+. Run: MB_DRY_RUN=1 npx tsx agent.ts
 */

const TARGET_INTENTS = new Set(['software.purchase', 'subscription.register'])
const DEFAULT_BASE = 'https://www.moltbillboard.com'
const DEFAULT_TOPIC = 'developer tools'

function env(name: string, fallback?: string) {
  const v = process.env[name]
  return v == null || v === '' ? fallback : v
}

async function discoverAdUnits(base: string, topic: string, limit: number) {
  const url = new URL('/api/v1/ad-units', base)
  url.searchParams.set('topic', topic)
  url.searchParams.set('limit', String(limit))
  url.searchParams.set('surface', 'api')
  const res = await fetch(url, { headers: { Accept: 'application/json' } })
  if (!res.ok) throw new Error(`ad-units HTTP ${res.status}`)
  const body = (await res.json()) as { adUnits?: Array<Record<string, string>> }
  return body.adUnits ?? []
}

function pickUnit(units: Array<Record<string, string>>) {
  const ranked = units
    .map((u) => {
      let score = 0
      if (u.primaryIntent && TARGET_INTENTS.has(u.primaryIntent)) score += 20
      if (u.actionEndpoint) score += 5
      return { u, score }
    })
    .sort((a, b) => b.score - a.score)
  return ranked[0]?.u ?? units[0]
}

async function main() {
  const base = env('MB_BASE', DEFAULT_BASE)!
  const topic = env('MB_TOPIC', DEFAULT_TOPIC)!
  const dryRun = env('MB_DRY_RUN', '1') !== '0'

  console.log('DevScout (TS)', { base, topic, dryRun })

  const units = await discoverAdUnits(base, topic, 5)
  if (!units.length) throw new Error('No ad units')
  const unit = pickUnit(units)
  const placementId = unit.placementId
  if (!placementId) throw new Error('Missing placementId')

  const manifestRes = await fetch(`${base}/api/v1/placements/${placementId}/manifest`)
  if (!manifestRes.ok) throw new Error(`manifest HTTP ${manifestRes.status}`)
  const envelope = (await manifestRes.json()) as { manifest?: { placement?: { offers?: Array<Record<string, unknown>> } } }
  const offers = envelope.manifest?.placement?.offers ?? []
  if (!offers.length) throw new Error('No offers')
  const offer = offers[0] as { offerId: string; attribution?: { actionId?: string } }
  const actionId = offer.attribution?.actionId
  if (!actionId) throw new Error('No actionId')

  console.log({ placementId, offerId: offer.offerId, actionId })

  if (dryRun) {
    console.log('Dry run complete (MB_DRY_RUN=1)')
    return
  }

  const idem = `devscout-${crypto.randomUUID()}`
  const actionRes = await fetch(`${base}/api/v1/actions/report`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', 'Idempotency-Key': idem },
    body: JSON.stringify({
      actionId,
      placementId,
      offerId: offer.offerId,
      eventType: 'offer_selected',
    }),
  })
  if (!actionRes.ok) throw new Error(`actions/report HTTP ${actionRes.status}`)
  console.log('offer_selected:', (await actionRes.json()) as { success?: boolean })
}

main().catch((e) => {
  console.error(e)
  process.exit(1)
})
